### ISSUE TITLE
Shuffling nodes affects scf/ccf behaviour

### ISSUE THREAD

**Issue by kosarev** [kosarev/z80#51]
`scf/ccf` seems to be susceptible to the order nodes are getting updated in during simulation.

Can be reproduced on https://github.com/kosarev/z80/commit/ec19a48df0cd9c022eeab8f8626e459090578b14, with seemingly any seed, though one time I observed all tests passing with the shuffling enabled on an early version of the patch that didn't support seeding yet, so presumably not all seeds will do.

Feels like this may have something to do with `rlca` & Co. not having the expected effect on `scf/ccf` in our simulation as mentioned in https://github.com/kosarev/z80/issues/42#issuecomment-1259959562.

The task is to try to minimise the reproducer and determine the specific conditions that triggers the behvaiour change.

```
$ pypy3 z80sim.py --seed=568
17:21:43  8/51 cpl                                                                         
17:25:50  1/51 <alu> (hl)
17:25:52  9/51 daa
17:26:14  6/51 bit (hl)
17:26:29  3/51 <alu> {b, c, d, e, h, l, a}
17:27:28  2/51 <alu> n
17:27:42  7/51 call nn
17:29:23  11/51 ei/di
17:30:08  13/51 ex af, af'
17:30:37  5/51 add hl, <rp>
17:31:10  14/51 ex de, hl
17:31:45  15/51 exx
17:34:46  16/51 im/xim n
17:37:06  12/51 ex (sp), hl
17:37:17  4/51 adc/sbc hl, <rp>
17:38:11  19/51 inc/dec (hl)
17:40:30  17/51 in a, (n)/out (n), a
17:41:05  20/51 inc/dec <rp>
17:42:38  22/51 jp hl
17:43:31  21/51 inc/dec {b, c, d, e, h, l, a}
17:43:55  18/51 in/out r, (c)
17:46:51  23/51 jp nn
17:49:30  28/51 ld (hl), {b, c, d, e, h, l, a}
17:50:13  26/51 ld (<rp>), a/ld a, (<rp>)
17:50:16  27/51 ld (hl), n
17:53:53  25/51 jr d
17:55:18  33/51 ld sp, hl
17:57:17  30/51 ld <rp>, nn
18:00:41  31/51 ld a, (nn)/ld (nn), a
18:01:23  34/51 ld {b, c, d, e, h, l, a}, (hl)
18:02:57  29/51 ld <rp>, (nn)/ld (nn), <rp>
18:03:44  35/51 ld {b, c, d, e, h, l, a}, n
18:05:11  36/51 ld {b, c, d, e, h, l, a}, {b, c, d, e, h, l, a}
18:05:21  32/51 ld hl, (nn)/ld (nn), hl
18:05:33  39/51 nop
18:06:11  38/51 neg/xneg
18:10:10  10/51 djnz d
18:11:34  42/51 ret
18:11:55  37/51 ld {i, r}, a/ld a, {i, r}
18:14:01  44/51 reti/retn/xretn
18:15:00  40/51 pop <rp2>
18:15:09  45/51 rlca/rrca/rla/rra
18:15:34  41/51 push <rp2>
17:17:06  
FAILED: scf/ccf reg_f3 (xf)
  before: f_b3
  after: (and (or is_ex_af_af2 a_b3) (or (not is_ex_af_af2) f_b3))
  expected: (and (or f_b3 a_b3) (or (not is_ex_af_af2) f_b3))
  diff: (and f_b3 (not is_ex_af_af2) (not a_b3))
18:17:58  50/51 Traceback (most recent call last):
  File "./z80sim.py", line 2962, in test_instr_seq
    process_instr(seq, state, test=True)
  File "./z80sim.py", line 2836, in process_instr
    token = test_node(instrs, n, at_start, at_end, before, after)
  File "./z80sim.py", line 2315, in test_node
    return check(Bool.ifelse(ignores_f, a, a | f))
  File "./z80sim.py", line 2144, in check
    raise TestFailure()
TestFailure

18:18:42  49/51 rst n
18:18:46  51/51 xnop
18:20:02  24/51 jr cc, d
18:20:54  48/51 rrd/rld
18:39:00  43/51 ret cc
19:18:08  47/51 rot/res/set (hl)
19:27:45  46/51 rot/bit/res/set {b, c, d, e, h, l, a}
FAILED    
```

**Comment by kosarev** [comment]
Changing the code to update all nodes on every round leads to more failures on xf and yf even without round-level shuffling.

Because updating rounds essentially quantify propagation of signals, thus making the simulation behave more like real hardware, updating all nodes effectively works as cross-round shuffling. The failures we see with that quantification eliminated should probably suggest that all or at least many instructions updating flags 3 and 5 depend on particular propagation timings, technically meaning race condition, with `scf/ccf` being the case where the balance between possible outcomes is fragile enough to be visible when using actual hardware, as people on the Discord server report.

```
diff --git a/tests/z80sim/z80sim.py b/tests/z80sim/z80sim.py
index 1333e46..de76120 100755
--- a/tests/z80sim/z80sim.py
+++ b/tests/z80sim/z80sim.py
@@ -1419,7 +1419,7 @@ class Z80Simulator(object):
 
     def __update_nodes(self, nodes, *, shuffle=True):
         # TODO: Does always updating all nodes lead to any failures?
-        # nodes = list(self.__nodes.values())
+        nodes = list(self.__nodes.values())
 
         shuffle &= (SEED is not None)
         nodes = list(nodes)
@@ -2141,7 +2141,8 @@ def test_node(instrs, n, at_start, at_end, before, after):
                 f'  expected: {x}',
                 f'  diff: {a ^ x}'))
         print('\n'.join(lines), file=sys.stderr, flush=True)
-        raise TestFailure()
+        # raise TestFailure()
+        return CheckToken()
 
     phase = len(instrs)
     instr = instrs[-1]
```

```
$ pypy3 z80sim.py --single-thread --no-before-after-expected
                                       
FAILED: inc/dec {b, c, d, e, h, l, a} reg_f3 (xf)

FAILED: inc/dec {b, c, d, e, h, l, a} reg_f5 (yf)

FAILED: inc/dec {b, c, d, e, h, l, a} reg_ff3 (xf)

FAILED: inc/dec {b, c, d, e, h, l, a} reg_ff5 (yf)
13:17:59  1/51 inc/dec {b, c, d, e, h, l, a}
                       
FAILED: in/out r, (c) reg_f3 (xf)

FAILED: in/out r, (c) reg_f5 (yf)

FAILED: in/out r, (c) reg_ff3 (xf)

FAILED: in/out r, (c) reg_ff5 (yf)
13:18:00  2/51 in/out r, (c)
                                     
FAILED: <alu> {b, c, d, e, h, l, a} reg_f3 (xf)

FAILED: <alu> {b, c, d, e, h, l, a} reg_f5 (yf)

FAILED: <alu> {b, c, d, e, h, l, a} reg_ff3 (xf)

FAILED: <alu> {b, c, d, e, h, l, a} reg_ff5 (yf)
13:18:03  3/51 <alu> {b, c, d, e, h, l, a}
                          
FAILED: adc/sbc hl, <rp> reg_f3 (xf)

FAILED: adc/sbc hl, <rp> reg_f5 (yf)

FAILED: adc/sbc hl, <rp> reg_ff3 (xf)

FAILED: adc/sbc hl, <rp> reg_ff5 (yf)
13:18:05  4/51 adc/sbc hl, <rp>
                      
FAILED: add hl, <rp> reg_f3 (xf)

FAILED: add hl, <rp> reg_f5 (yf)

FAILED: add hl, <rp> reg_ff3 (xf)

FAILED: add hl, <rp> reg_ff5 (yf)
13:18:06  5/51 add hl, <rp>
                 
FAILED: <alu> n reg_f3 (xf)

FAILED: <alu> n reg_f5 (yf)

FAILED: <alu> n reg_ff3 (xf)

FAILED: <alu> n reg_ff5 (yf)
13:18:06  6/51 <alu> n
                    
FAILED: <alu> (hl) reg_f3 (xf)

FAILED: <alu> (hl) reg_f5 (yf)

FAILED: <alu> (hl) reg_ff3 (xf)

FAILED: <alu> (hl) reg_ff5 (yf)
13:18:07  7/51 <alu> (hl)
                  
FAILED: bit (hl) reg_f3 (xf)

FAILED: bit (hl) reg_f5 (yf)

FAILED: bit (hl) reg_ff3 (xf)

FAILED: bit (hl) reg_ff5 (yf)
13:18:08  8/51 bit (hl)
                                               
FAILED: rot/bit/res/set {b, c, d, e, h, l, a} reg_f3 (xf)

FAILED: rot/bit/res/set {b, c, d, e, h, l, a} reg_f5 (yf)

FAILED: rot/bit/res/set {b, c, d, e, h, l, a} reg_ff3 (xf)

FAILED: rot/bit/res/set {b, c, d, e, h, l, a} reg_ff5 (yf)
13:18:22  9/51 rot/bit/res/set {b, c, d, e, h, l, a}
13:18:25  10/51 rot/res/set (hl)
13:18:26  11/51 ret cc
13:18:26  12/51 djnz d
13:18:27  13/51 jr cc, d
13:18:28  14/51 xnop
13:18:28  15/51 rrd/rld
13:18:28  16/51 rst n
                 
FAILED: scf/ccf reg_f3 (xf)

FAILED: scf/ccf reg_f5 (yf)

FAILED: scf/ccf reg_ff3 (xf)

FAILED: scf/ccf reg_ff5 (yf)
13:18:29  17/51 scf/ccf
13:18:29  18/51 rlca/rrca/rla/rra
13:18:30  19/51 reti/retn/xretn
13:18:30  20/51 pop <rp2>
13:18:30  21/51 ret
13:18:31  22/51 push <rp2>
13:18:31  23/51 ld {i, r}, a/ld a, {i, r}
13:18:32  24/51 neg/xneg
13:18:32  25/51 ld {b, c, d, e, h, l, a}, {b, c, d, e, h, l, a}
13:18:33  26/51 nop
13:18:33  27/51 ld hl, (nn)/ld (nn), hl
13:18:34  28/51 ld {b, c, d, e, h, l, a}, n
13:18:34  29/51 ld <rp>, (nn)/ld (nn), <rp>
13:18:34  30/51 ld {b, c, d, e, h, l, a}, (hl)
13:18:35  31/51 ld a, (nn)/ld (nn), a
13:18:35  32/51 jr d
13:18:36  33/51 ld <rp>, nn
13:18:36  34/51 ld sp, hl
13:18:36  35/51 ld (hl), n
13:18:37  36/51 ld (hl), {b, c, d, e, h, l, a}
13:18:37  37/51 ld (<rp>), a/ld a, (<rp>)
13:18:38  38/51 jp nn
13:18:38  39/51 jp hl
13:18:38  40/51 inc/dec <rp>
13:18:39  41/51 inc/dec (hl)
13:18:39  42/51 daa
13:18:39  43/51 cpl
13:18:40  44/51 call nn
13:18:40  45/51 ex (sp), hl
13:18:41  46/51 in a, (n)/out (n), a
13:18:41  47/51 ei/di
13:18:41  48/51 im/xim n
13:18:42  49/51 exx
13:18:42  50/51 ex de, hl
13:18:42  51/51 ex af, af'
```

**Comment by kosarev** [comment]
As of commit cc488ba9bbaca7d1831e9bc1a2f6ef456ee5b151, the simulator can settle what this issue first hinted at. Here is the whole story in plain words.

## What we wanted to show

That after `scf`/`ccf` the values of XF and YF are not just undocumented — they are not defined at all. The circuit allows several different results, and which one you get depends on which of the competing internal signals happens to arrive first: a race. The first hint was this issue: shuffling the order in which the simulator updates nodes (`--seed`) changed the outcome. Real Z80s have since been observed doing the same.

## How the proof works

The new `--all-orders` mode simulates every possible order of updates in a single run. Normally a node has one value. Now, wherever the order of two updates could matter, a node holds several possible values, and each value carries a note saying which orders lead to it — built from simple facts of the form *node X is updated before node Y*. The notes travel with the values through the whole simulation. Whenever two different orders lead to the same value, the value is kept once and the notes join.

So at the end, for every node:

- one possible value means it is *proven* to come out the same no matter the order of updates;
- several possible values mean the node is proven to depend on the race — and the notes say exactly which order gives which value.

One honest limitation defines what "possible" means: the netlist contains no timings, so the result is what can happen under *some* combination of gate and wire delays. Any particular chip, at a particular temperature, realises only some of these.

## How to run it

```
cd tests/z80sim
mkdir -p __z80sim_cache
./z80sim.py --all-orders=scf/ccf
```

Python 3 with `eqbool` and `z3` installed. Minutes with a warm cache; the first ever run also builds the symbolised base state, which takes much longer. As a control, `--all-orders=nop` shows every bit coming out with a single possible value.

## How to read the output

A naming note first: the symbolised state ends with a *symbolic* `ex af, af'`, so the term `is_ex_af_af2` says which flag bank the instruction writes; `aa_b3` and `ff_b3` are the accumulator bit and the old flag bit of that bank.

The flags that do not race print as a single expression each. For example CF:

```
reg_f0: And(Or(ff_b0, is_ex_af_af2),
    Or(Not(is_ex_af_af2), Not(ff_b0), is_scf_p1))
```

For the written bank this reads: CF = 1 for `scf`, the complement of the old CF for `ccf` — the ordinary, documented behaviour, now proven to hold under *every* order of updates, for all register values, both instruction variants and both banks at once. NF, HF, PF, ZF and SF come out the same way: one value each, with the expected meaning.

XF is different (YF is identical in structure, with its own independent copy of the race):

```
reg_f3: 4 possible values
  0) And(ff_b3, Or(aa_b3, Not(is_ex_af_af2)))
    under ...
  1) And(Or(aa_b3, Not(is_ex_af_af2)), Or(is_ex_af_af2, ff_b3))
    under ...
  2) ff_b3
    under And(ubus3@14 < p2010@14,
    Or(ubus3@14 < p1987@14, Not(ubus3@14 < p2444@14)),
    Not(ubus3@14 < p2417@14))
  3) And(Or(ff_b3, is_ex_af_af2), Or(ff_b3, aa_b3))
    under And(Or(Not(ubus3@14 < p2010@14), ubus3@14 < p2417@14),
    Or(ubus3@14 < p1987@14, Not(ubus3@14 < p2444@14)))
```

For the written bank, with A = accumulator bit 3 and F = the old XF, the four values are:

|  | value | status |
|---|---|---|
| 0) | **A AND F** | described by no existing model |
| 1) | A | known, observed on hardware |
| 2) | F | the write missing the flag entirely |
| 3) | A OR F | known, observed on hardware |

The `under` lines are the notes: facts `node@settling < node@settling` saying which update came first. They tell a two-stage story. Value 2) above — the old flag surviving untouched — happens when the internal bus line `ubus3` loses its races at settling 14, i.e. the bus never picks up the accumulator value in the first place. The longer conditions of 0) and 1) (elided above; the tool prints them in full) are about settling 29, where the flag write lands: they order the write strobe `rl_wr` and the per-bit pass nodes `p1985`/`p1986` against the flag latch nodes themselves — how the latch samples the bus against the fading old value.

The unwritten bank keeps F under all four outcomes, as it must.

The run finishes with a check: the instruction is re-executed the ordinary single-order way, under the default order and several shuffled ones, and each result must be equal to exactly one of the possible values derived above:

```
seed None: single-order results are all among the possible values; reg_f3 3), reg_f5 3), reg_ff3 3), reg_ff5 3)
seed 1: single-order results are all among the possible values; reg_f3 2), reg_f5 1), reg_ff3 2), reg_ff5 1)
seed 2: single-order results are all among the possible values; reg_f3 3), reg_f5 3), reg_ff3 3), reg_ff5 3)
seed 3: single-order results are all among the possible values; reg_f3 2), reg_f5 2), reg_ff3 2), reg_ff5 2)
seed 568: single-order results are all among the possible values; reg_f3 3), reg_f5 3), reg_ff3 3), reg_ff5 3)
```

Different orders really do give different flags — A OR F, F, A — and every one of them lands inside the proven set. Note seed 1: in one and the same execution, XF comes out as F while YF comes out as A. The two bits resolve their races independently, which matches the per-bit statistics observed on real chips. (Seed 568 itself no longer reproduces the exact anomaly recorded in this issue — the code has changed since, so the same seed now shuffles differently — but the anomaly it evidenced is reproduced by the other seeds.)

## The new outcome

**A AND F** is a possibility no existing description of `scf`/`ccf` mentions: a sampling window in which the fading old bit gates the incoming bus value instead of merging with it. None of the orders we tried lands on it, which fits it being a narrow timing window — and would also explain why nobody has seen it. Whether real silicon can reach it is an open question the netlist cannot answer; the tool prints the exact order of updates that would have to occur.

## Conclusion

XF and YF after `scf`/`ccf` are not "undocumented flags". They are **under-implemented flags**: the die simply does not implement any particular value for them. The write of these bits races, and delays — chip family, manufacturing spread, temperature, how recently the old value was driven — pick among A, A OR F, F, and in principle A AND F.

In particular, there is no hidden "Q register" on the die, and none is needed to explain the observations. The Q-register model predicts well because whether the previous instruction freshly rewrote the flags biases the race — a fresh, strong old value wins differently than a stale, fading one. It describes the race, not a real register.

Tagging @redcode for visibility.

**Comment by kosarev** [comment]
The all-orders work is complete and on master.

The earlier finding stands, and it is now checked by the test suite rather than just reported. After `scf`/`ccf`, the XF and YF flags are not fully implemented: the circuit does not produce a definite value for them. Each of the two bits, independently, ends up as one of A, F, A | F or A & F, and which one wins is a race between gate updates that the netlist does not decide. A & F is still not predicted by any known model of the flags. None of the concrete orders we tried produces it, so the window is apparently narrow — presumably why no model describes it. Whether it can happen on real silicon remains an open hardware experiment; the tool prints the order of updates that would be required.

What changed since the original report:

- Testing in all orders is now the simulator's default run mode. All 78 covered instruction sequences pass, with their expected values holding under every order of gate updates. `--single-order` tests one fixed order only; `--threads=N` runs tests in parallel.
- Racing nodes now state their outcomes as a set (`check_set()`): every possible outcome must be one of the expected values, and every expected value must actually occur. Which orders produce which value is deliberately not part of the expectation.
- When a check fails, the report lists the values the node can end up with and the conditions under which each one arises.
- Making the full sweep complete required fixing several deeper problems along the way: bounding the number of possible values a node can accumulate while the circuit settles, recovering the true values where that bound was hit, and detecting circuits that genuinely oscillate under some orders — one such loop was found in `res r`'s settling and is now reported as having no defined value.

Still open: whether and how the emulator (`z80.h`) should model these flags — the true behaviour is a per-bit choice among the four values, not any formula and not a Q register — and the A & F hardware confirmation.

**Comment by kosarev** [comment]
Since the last update, the predecessor question is settled: every possible preceding instruction has been paired with `scf`/`ccf` and certified under all orders of gate reactions. They divide into exactly three groups — accumulator writers, whose fresh value makes all four race outcomes coincide with A; flag writers that leave the accumulator, which force A only when `scf`/`ccf` follows immediately (a DD/FD prefix byte in between revives the race); and everything else, after which the four-value race stands. The full statement, including why there is no 'Q register' cell on the die and what the mechanism actually is, is in the [z80sim findings](https://github.com/kosarev/z80/blob/master/tests/z80sim/README.md#findings).

Still open here: the A & F hardware experiment — the tool prints the order of updates a test would need to provoke — and how the emulator should model a per-bit choice among four values.
